File: android/src/main/java/com/margelo/nitro/nitrospeech/recognizer/AutoStopper.kt
package com.margelo.nitro.nitrospeech.recognizer

import android.os.Handler
import android.os.Looper
import kotlin.math.max

class AutoStopper(
    silenceThresholdMs: Double?,
    progressIntervalMs: Double?,
    private val onProgress: (Double) -> Unit,
    val onTimeout: () -> Unit,
) {
    companion object {
        private const val DEFAULT_SILENCE_THRESHOLD_MS = 8000.0
        private const val DEFAULT_PROGRESS_INTERVAL_MS = 1000.0
        private const val MIN_PROGRESS_INTERVAL_MS = 50.0
    }

    private val logger = Logger(disable = false)

    private var silenceThresholdMs: Double = clampMs(silenceThresholdMs ?: DEFAULT_SILENCE_THRESHOLD_MS)
    private var progressIntervalMs: Double = clampMs(progressIntervalMs ?: DEFAULT_PROGRESS_INTERVAL_MS)

    private var timeLeftMs: Double = this.silenceThresholdMs
    private var isStopped = false
    private var didTimeout = false
    private var isTimerScheduled = false

    private val handler = Handler(Looper.getMainLooper())

    private val tickRunnable = Runnable { tick() }

    fun resetTimer() {
        logger.log("resetTimer | isStopped: $isStopped | ms: ${System.currentTimeMillis()}")
        handler.removeCallbacks(tickRunnable)
        isTimerScheduled = false
        if (isStopped) return
        didTimeout = false
        timeLeftMs = silenceThresholdMs
        if (timeLeftMs > 0) {
            onProgress(timeLeftMs)
        }
        scheduleNextTickLocked()
    }

    fun stop() {
        isStopped = true
        handler.removeCallbacks(tickRunnable)
        isTimerScheduled = false
    }

    fun updateSilenceThreshold(newThresholdMs: Double) {
        silenceThresholdMs = clampMs(newThresholdMs)
    }

    fun addMsOnce(extraMs: Double) {
        if (isStopped || !extraMs.isFinite()) return
        logger.log("addMsOnce | extraMs: $extraMs")
        timeLeftMs += extraMs
        didTimeout = false
        if (timeLeftMs > 0 && isTimerScheduled) {
            onProgress(timeLeftMs)
        }
    }

    fun updateProgressInterval(newIntervalMs: Double) {
        if (isStopped) return
        logger.log("updateProgressInterval | newIntervalMs: $newIntervalMs")
        progressIntervalMs = clampMs(newIntervalMs)
        if (isTimerScheduled) {
            scheduleNextTickLocked()
        }
    }

    private fun scheduleNextTickLocked() {
        handler.removeCallbacks(tickRunnable)
        val delayMs = progressIntervalMs.toLong().coerceAtLeast(MIN_PROGRESS_INTERVAL_MS.toLong())
        handler.postDelayed(tickRunnable, delayMs)
        isTimerScheduled = true
    }

    private fun tick() {
        if (isStopped || didTimeout) return
        timeLeftMs -= progressIntervalMs
        if (timeLeftMs > 0) {
            logger.log("onProgress | timeLeftMs: $timeLeftMs")
            onProgress(timeLeftMs)
            scheduleNextTickLocked()
            return
        }
        timeLeftMs = 0.0
        didTimeout = true
        handler.removeCallbacks(tickRunnable)
        isTimerScheduled = false
        logger.log("onTimeout | ms: ${System.currentTimeMillis()}")
        onTimeout()
    }

    private fun clampMs(value: Double): Double {
        if (!value.isFinite()) return MIN_PROGRESS_INTERVAL_MS
        return max(MIN_PROGRESS_INTERVAL_MS, value)
    }
}

File: ios/Shared/AutoStopper.swift
import Foundation

final class AutoStopper {
    private static let defaultSilenceThresholdMs = 8000.0
    private static let defaultProgressIntervalMs = 1000.0
    private static let minProgressIntervalMs = 50.0

    private let lg = Lg(prefix: "AutoStopper", disable: true)
    
    private let queue = DispatchQueue(label: "com.margelo.nitrospeech.autostopper")

    private var silenceThresholdMs: Double
    private var progressIntervalMs: Double
    private var timeLeftMs: Double
    private var isStopped = false
    private var didTimeout = false
    private var timer: DispatchSourceTimer?

    private let onProgress: (Double) -> Void
    private let onTimeout: () -> Void

    init(
        silenceThresholdMs: Double?,
        progressIntervalMs: Double?,
        onProgress: @escaping (Double) -> Void,
        onTimeout: @escaping () -> Void
    ) {
        let threshold = Self.clampMs(silenceThresholdMs ?? Self.defaultSilenceThresholdMs)
        let interval = Self.clampMs(progressIntervalMs ?? Self.defaultProgressIntervalMs)
        self.silenceThresholdMs = threshold
        self.progressIntervalMs = interval
        self.timeLeftMs = threshold
        self.onProgress = onProgress
        self.onTimeout = onTimeout
    }

    deinit {
        // Never `queue.sync` here. Every method dispatches with
        // `queue.async { guard let self ... }`, which creates a strong
        // reference that is released when the block ends, i.e. ON `queue`.
        // Once the owner has released its own reference (RecognizerEngine's
        // cleanup calls deinitAutoStop on every stop), that block-local
        // reference is the last one, so deinit runs on `queue` and a sync onto
        // the current serial queue deadlocks: libdispatch traps in
        // __DISPATCH_WAIT_FOR_QUEUE__ and the app dies with SIGTRAP.
        //
        // No synchronisation is needed instead: deinit means no other
        // reference exists, and the timer's event handler is [weak self].
        isStopped = true
        timer?.cancel()
        timer = nil
    }

    func resetTimer(from: String) {
        queue.async { [weak self] in
            guard let self, !self.isStopped else { return }
            lg.log("[resetTimer] from:\(from)")
            self.didTimeout = false
            self.timeLeftMs = self.silenceThresholdMs
            self.startOrRescheduleTimerLocked()
            if self.timeLeftMs > 0 {
                self.onProgress(self.timeLeftMs)
            }
        }
    }

    func updateThreshold(_ newThresholdMs: Double, from: String) {
        queue.async { [weak self] in
            guard let self, !self.isStopped else { return }
            lg.log("[updateThreshold] from:\(from) newThresholdMs:\(newThresholdMs)")
            self.silenceThresholdMs = Self.clampMs(newThresholdMs)
        }
    }

    func addMsOnce(_ extraMs: Double, from: String) {
        queue.async { [weak self] in
            guard let self, !self.isStopped, extraMs.isFinite else { return }
            lg.log("[addMsOnce] from:\(from) extraMs:\(extraMs)")
            self.timeLeftMs += extraMs
            self.didTimeout = false
            if self.timeLeftMs > 0, self.timer != nil {
                self.onProgress(self.timeLeftMs)
            }
        }
    }

    func updateProgressInterval(_ newIntervalMs: Double, from: String) {
        queue.async { [weak self] in
            guard let self, !self.isStopped else { return }
            lg.log("[updateProgressInterval] from:\(from) newIntervalMs:\(newIntervalMs)")
            self.progressIntervalMs = Self.clampMs(newIntervalMs)
            if self.timer != nil {
                self.startOrRescheduleTimerLocked()
            }
        }
    }

    func stop() {
        queue.async { [weak self] in
            guard let self else { return }
            self.stopLocked()
            self.timeLeftMs = 0
        }
    }

    private func startOrRescheduleTimerLocked() {
        timer?.cancel()
        timer = nil

        let source = DispatchSource.makeTimerSource(queue: queue)
        let intervalNs = UInt64(progressIntervalMs * 1_000_000)
        source.schedule(
            deadline: .now() + .nanoseconds(Int(intervalNs)),
            repeating: .nanoseconds(Int(intervalNs))
        )
        source.setEventHandler { [weak self] in
            self?.tickLocked()
        }
        timer = source
        source.resume()
    }

    private func tickLocked() {
        guard !isStopped else { return }
        guard !didTimeout else { return }

        timeLeftMs -= progressIntervalMs
        if timeLeftMs > 0 {
            lg.log("[onProgress] timeLeftMs:\(timeLeftMs)")
            onProgress(timeLeftMs)
            return
        }

        timeLeftMs = 0
        didTimeout = true
        cancelTimerLocked()
        lg.log("[onTimeout]")
        onTimeout()
    }

    private func stopLocked() {
        isStopped = true
        cancelTimerLocked()
    }

    private func cancelTimerLocked() {
        timer?.cancel()
        timer = nil
    }

    private static func clampMs(_ value: Double) -> Double {
        if !value.isFinite {
            return minProgressIntervalMs
        }
        return max(minProgressIntervalMs, value)
    }
}
